src/gf.c

```diff
@@ -2526,7 +2534,45 @@ jl_code_instance_t *jl_compile_method_internal(jl_method_instance_t *mi, size_t
         }
     }
 
-    codeinst = jl_generate_fptr(mi, world);
+    // Ok, compilation is enabled. We'll need to try to compile something (probably).
+    // Try to find a codeinst we have already inferred (e.g. while we were compiling
+    // something else).
+    codeinst = jl_method_inferred_with_source(mi, world);
+
+    // Everything from here on our is considered (user facing) compile time
+    uint64_t start = jl_typeinf_timing_begin();
+    int is_recompile = jl_atomic_load_relaxed(&mi->cache) != NULL;
+
+    // This codeinst hasn't been previously inferred do that now
+    if (!codeinst) {
+        // Don't bother inferring toplevel thunks or macros - the performance cost of inference is likely
+        // to significantly exceed the actual runtime.
+        int should_skip_inference = !jl_is_method(mi->def.method) || jl_symbol_name(mi->def.method->name)[0] == '@';
+
+        if (!should_skip_inference) {
+            codeinst = jl_type_infer(mi, world, 0, SOURCE_MODE_ABI);
+        }
+    }
+
+    if (codeinst) {
+        // Something went very wrong. Bail out.
+        if (jl_atomic_load_acquire(&codeinst->invoke) != NULL) {
+            jl_typeinf_timing_end(start, is_recompile);
+            // Already compiled - e.g. constabi, or compiled by a different thread while we were waiting.
+            return codeinst;
+        }
+
+        JL_GC_PUSH1(&codeinst);
+        jl_compile_codeinst(codeinst);
```

Comment:
```suggestion
        codeinst = jl_compile_codeinst(codeinst);
```

IIRC, codegen sometimes either can't satify this, or realizes it would be a waste of time to do so, and should substitue in a better result here (e.g. one that is already compiled, but had to wait for the lock to check)

Reply:
Failure here is indicated by not setting ->invoke. I agree that it would be preferable to re-check the cache under the codegen lock, but I think that needs to happen at the caller side.

Reply:
But failure here does not result in leaving invoke NULL, but rather in compiling an unspecialized copy. It seems awkward to be changing the meaning of "failure" in codegen here to mean that multithreaded races where code is required on multiple threads at once will JIT one good copy for one thread and then run unspecialized copies on all of the other threads that lost this race